File: day6/day6.py
# -*- coding: utf-8 -*-

### DAY 6

## PART 1

signal = list()

with open(r"data.txt", "r") as f:
    for line in f:
        for letter in line:
            signal.append(letter)

def findMarker(markerLength):

    firstMarker = 0

    i = markerLength
    while(firstMarker == 0):
        if len(set(signal[(i-markerLength):i])) == markerLength:
            firstMarker = i
        else:
            i += 1

    return(firstMarker)

markerLength = 4
print(f"{findMarker(markerLength)} characters need to be processed before the first start-of-packet marker.")


## PART 2

markerLength = 14
print(f"{findMarker(markerLength)} characters need to be processed before the first start-of-packet marker.")

File: day6/day6.R

### DAY 6

## PART 1

signal <- strsplit(readLines("data.txt"), split = "")[[1]]

find.marker <- function(marker.length) {
  
  first.marker <- 0
  
  i <- marker.length
  while (first.marker == 0) {
    
    if (length(unique(signal[(i-marker.length+1):i])) == marker.length) {
      first.marker <- i
    } else {
      i <- i + 1
    }
  }
  
  return(first.marker)
}

cat(find.marker(4),
    " characters need to be processed before the first start-of-packet marker.\n",
    sep = "")


## PART 2

cat(find.marker(14),
    " characters need to be processed before the first start-of-packet marker.\n",
    sep = "")
